### PR TITLE
Line Segment Kernel to Alpaka

### PR DESCRIPTION
**Pull request by GNiendorf**
Moves the LS kernel to Alpaka.

Timing
Screenshot 2023-03-10 at 11 30 45 PM

Validation Plots: [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR251/PR251_8d759a-PU200/summary/)

### CODE REVIEW

**Review by VourMa — COMMENTED** [review]
The PR looks good, just a couple of comments to understand.

Not specifically for this PR but maybe for the next ones, could you make the comparison plots instead of the raw ones (there should be this possibility in the new plotting scripts)? It's difficult to judge if the performance changes otherwise.

**Comment by sgnoohc** [comment]
With the new script run something like

python lst_performance_plot.py —compare —comp_labels master,PR251 histogram_master.root histogram_PR251.root

I am typing this on my cell phone.
Sorry for the typos and bad double dash formatting… but i think the gist is there.
Use the help option -h for more info.

—
Dr. P. Chang, Asst. Prof., Univ. of Florida

On Mar 18, 2023, at 9:02 AM, Manos Vourliotis ***@***.***> wrote:

﻿
[External Email]

@VourMa commented on this pull request.

The PR looks good, just a couple of comments to understand.

Not specifically for this PR but maybe for the next ones, could you make the comparison plots instead of the raw ones (there should be this possibility in the new plotting scripts)? It's difficult to judge if the performance changes otherwise.

________________________________

In SDL/Segment.cu<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSegmentLinking%2FTrackLooper%2Fpull%2F251%23discussion_r1141011220&data=05%7C01%7Cp.chang%40ufl.edu%7Ca8a6cff529534c668b0f08db27b10235%7C0d4da0f84a314d76ace60a62331e1b84%7C0%7C0%7C638147413419565317%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=LIJI6sPUT0rdpfa4GBK7iT58qgssrTU8EXFuaGBDDKM%3D&reserved=0>:

> @@ -29,8 +25,7 @@ void SDL::segments::resetMemory(unsigned int nMemoryLocationsx, unsigned int nLo
     cudaMemsetAsync(pLSHitsIdxs, 0,maxPixelSegments * sizeof(uint4),stream);
 }

-
-void SDL::createSegmentArrayRanges(struct modules& modulesInGPU, struct objectRanges& rangesInGPU, struct miniDoublets& mdsInGPU, uint16_t& nLowerModules, unsigned int& nTotalSegments, cudaStream_t stream, const uint16_t& maxPixelSegments)
+void SDL::createSegmentArrayRanges(struct modules& modulesInGPU, struct objectRanges& rangesInGPU, struct miniDoublets& mdsInGPU, uint16_t& nLowerModules, int& nTotalSegments, cudaStream_t stream, const uint16_t& maxPixelSegments)


I see that this function takes as an argument a cudaStream_t. Shouldn't this be changed to a corresponding Alpaka object?

I can see that this is true for more functions below. What's our strategy for dealing with those?

________________________________

In SDL/Segment.cuh<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSegmentLinking%2FTrackLooper%2Fpull%2F251%23discussion_r1141011792&data=05%7C01%7Cp.chang%40ufl.edu%7Ca8a6cff529534c668b0f08db27b10235%7C0d4da0f84a314d76ace60a62331e1b84%7C0%7C0%7C638147413419565317%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=WJJn86kIhIDxe6wdP%2BpjJSe2Kz38yv2OLgT6TiyIFEg%3D&reserved=0>:

> @@ -2,15 +2,14 @@
 #define Segment_cuh

 #include "Constants.cuh"
+#include "Constants.h"


Partially unrelated to this PR but is there a point to have twoConstants files?

File: SDL/Segment.cu [hunk 1]
```diff
@@ -29,8 +25,7 @@ void SDL::segments::resetMemory(unsigned int nMemoryLocationsx, unsigned int nLo
     cudaMemsetAsync(pLSHitsIdxs, 0,maxPixelSegments * sizeof(uint4),stream);
 }
 
-
-void SDL::createSegmentArrayRanges(struct modules& modulesInGPU, struct objectRanges& rangesInGPU, struct miniDoublets& mdsInGPU, uint16_t& nLowerModules, unsigned int& nTotalSegments, cudaStream_t stream, const uint16_t& maxPixelSegments)
+void SDL::createSegmentArrayRanges(struct modules& modulesInGPU, struct objectRanges& rangesInGPU, struct miniDoublets& mdsInGPU, uint16_t& nLowerModules, int& nTotalSegments, cudaStream_t stream, const uint16_t& maxPixelSegments)
```
[the remark this answers is not in the document]
Comment (GNiendorf) [reply SDL/Segment.cu:28]: That will be addressed in a later PR after all the kernels are moved over. I'm not sure yet what the best solution is yet, but we'll figure it out.

File: SDL/Segment.cuh [hunk 2]
```diff
@@ -2,15 +2,14 @@
 #define Segment_cuh
 
 #include "Constants.cuh"
+#include "Constants.h"
```
[the remark this answers is not in the document]
Comment (GNiendorf) [reply SDL/Segment.cuh:5]: Good question. I think the variables in Constants.h are defined on the CPU, and Constants.cuh are for constants that live on the GPU, but we can probably combine these two files after we move fully to alpaka.

**Review by VourMa — APPROVED** [review]
Thanks for the replies to the comments, they make sense. Let's keep them in mind for the future. Merging this...

**Pull request merged**